jobserv/models.py

```diff
@@ -458,45 +458,62 @@ def pop_queued(worker):
         cursor = conn.cursor()
 
         rows = cursor.execute('''
-            SELECT runs.id FROM runs
+            SELECT
+              runs.id, runs.build_id, runs._status,
+              projects.id, projects.synchronous_builds
+            FROM runs
             JOIN builds on builds.id = runs.build_id
             JOIN projects on projects.id = builds.proj_id
             WHERE
-                runs._status = 1
+                runs._status in (1, 2)
               AND ({tags})
-              AND
-                projects.id NOT IN (
-                    SELECT projects.id
-                    FROM projects
-                    JOIN builds ON projects.id = builds.proj_id
-                    JOIN runs on runs.build_id = builds.id
-                    WHERE
-                        runs._status = 2 AND projects.synchronous_builds = true
-                    GROUP BY
-                        projects.id, projects.name, projects.synchronous_builds
-                )
-              ORDER BY runs.queue_priority DESC, runs.build_id, runs.id ASC
-              LIMIT 1
+              ORDER BY
+                runs._status DESC, runs.queue_priority DESC,
+                runs.build_id ASC, runs.id ASC
             '''.format(tags=tags))
-        if rows:
-            # We have a suitable run
-            rows = cursor.execute('''
-                UPDATE runs
-                SET
-                    -- A trick to allow us to find the ID of the row we updated
-                    id = @run_id := id,
-                    _status = 2,
-                    worker_name = "{worker}"
-                WHERE
-                    id = {run_id}
-                '''.format(worker=worker.name, run_id=cursor.fetchone()[0]))
+
+        # By ordering the query above by Run._status, we'll get the active
+        # runs first so that we can build up this list of build ids that are
+        # active for synchronous projects runs with these build ids are okay
+        # to schedule
+        sync_projects = {}
+        okay_sync_builds = {}
+        for (run_id, build_id, status, proj_id, sync) in cursor.fetchall():
+            if status == 2 and sync:
```

Comment:
This is pretty slick.  Sort the RUNNING builds first and make a big list of them.  Then as the QUEUED builds get examined it's easy to see if there are conflicts.  This drops the inner query entirely.